stl/inc/format

```diff
@@ -2370,24 +2370,43 @@ _NODISCARD _OutputIt _Fmt_write(
     auto _Separators       = 0;
     string _Groups;
 
-    if ((_Specs._Alt || _Specs._Localized) && _Is_finite) {
-        for (auto _It = _Buffer_start; _It < _Result.ptr; ++_It) {
-            if (*_It == '.') {
-                _Radix_point = _It;
-            } else if (*_It == _Exponent) {
-                _Exponent_start = _It;
+    if (_Is_finite) {
+        if (_Specs._Alt || _Specs._Localized) {
+            for (auto _It = _Buffer_start; _It < _Result.ptr; ++_It) {
+                if (*_It == '.') {
+                    _Radix_point = _It;
+                } else if (*_It == _Exponent) {
+                    _Exponent_start = _It;
+                }
             }
-        }
-        _Integral_end = (_STD min) (_Radix_point, _Exponent_start);
+            _Integral_end = (_STD min) (_Radix_point, _Exponent_start);
 
-        if (_Specs._Alt) {
-            if (_Radix_point == _Result.ptr) {
+            if (_Specs._Alt && _Radix_point == _Result.ptr) {
                 // TRANSITION, decimal point may be wider
                 ++_Width;
                 _Append_decimal = true;
             }
 
-            if (_Specs._Type == 'g' || _Specs._Type == 'G') {
+            if (_Specs._Localized) {
+                _Groups     = _STD use_facet<numpunct<_CharT>>(_Locale._Get()).grouping();
+                _Separators = _Count_separators(_Integral_end - _Buffer_start, _Groups);
+            }
+        }
+
+        switch (_Format) {
+        case chars_format::hex:
+        case chars_format::scientific:
+            if (_Extra_precision != 0) {
+                while (*_Exponent_start != _Exponent) { // Trailing zeroes are in front of the exponent
+                    --_Exponent_start;
+                }
+            }
+            [[fallthrough]];
+        case chars_format::fixed:
+            _Zeroes_to_append = _Extra_precision;
+            break;
+        case chars_format::general:
+            if (_Specs._Alt) {

```

Comment:
Looking at it again, I noticed that `_Format` is initialized with `chars_format::general` which means that `_Format` is also `chars_format::general` for `_Specs._Type == '\0'`, so we should probably change this condition to:
```suggestion
            if (_Specs._Alt && _Specs._Type != '\0') {
```

Reply:
According to https://eel.is/c++draft/tab:format.type.float we should treat `none` as `general`

So I do not really see a reason to special case it here. Or the table is incomplete

Reply:
According to https://eel.is/c++draft/format.string.std#6.sentence-6 says `g` and `G` types handle `#` option differently than other types.

Reply:
It would still need to account for additional requested precision: That comment only talks about trimming zeroes not appending them

```cpp
        string_view expected{buffer, to_chars(begin(buffer), end(buffer), value, chars_format::general, 2000).ptr};
        assert(format("{:.2000}", value) == expected);
```

Reply:
I am not really an expert on floating point formatting. Does someone know of a test case where this actually happens?

Reply:
> That comment only talks about trimming zeroes not appending them

IIUC if you trim zeroes it means that you don't append them, and if you don't trim zeroes it means that you need to append them.

Reply:
I am not sure about that. This whole block of code is only ever needed if the require precision is *larger* than the number of digits needed to represent a floating point number. So we should only ever end up here if the user requested additional zeros. It has nothing to do with trimming.

Reply:
``` cpp
assert(format("{:#.2000}", 1.0) == "1.");
assert(format("{:#.2000g}", 1.0) == "1." + string(1999, '0'));
```

It appears that these input cases are missing from the current tests.

Reply:
Urgh, those a breaking :(

Reply:
I believe this one is wrong `assert(format("{:#.2000}", 1.0) == "1.");`

The `#` only requires that there is a dot even if it is not needed as far as I can tell. So this should be "1.0000...."

Reply:
`to_chars(first, last, 1.0, chars_format::general, 2000)` is `"1"`. `#` requires a decimal separator, so `"1."`.

Reply:
Resolving as it appears that this is covered by @miscco's latest added tests - please meow if you disagree. :smiley_cat:

Reply:
@StephanTLavavej 🐱📢
```
C:\Users\He\source\repos\STL\out>type gh2157.cpp
```
```cpp
#include <cstdio>
#include <format>
using namespace std;

int main() {
    puts(format("format(\"{{:#}}\", 1.0)    = {:#}", 1.0).c_str());
    puts(format("format(\"{{:#.6}}\", 1.0)  = {:#.6}", 1.0).c_str());
    puts(format("format(\"{{:#g}}\", 1.0)   = {:#g}", 1.0).c_str());
    puts(format("format(\"{{:#.6g}}\", 1.0) = {:#.6g}", 1.0).c_str());
    return 0;
}
```
```
C:\Users\He\source\repos\STL\out>cl /EHsc /W4 /WX /std:c++latest gh2157.cpp
用于 x64 的 Microsoft (R) C/C++ 优化编译器 19.30.30423 版
版权所有(C) Microsoft Corporation。保留所有权利。

/std:c++latest 作为最新的 C++
working 草稿中的语言功能预览提供。我们希望你提供有关 bug 和改进建议的反馈。
但是，请注意，这些功能按原样提供，没有支持，并且会随着工作草稿的变化
而更改或移除。有关详细信息，请参阅
https://go.microsoft.com/fwlink/?linkid=2045807。

gh2157.cpp
Microsoft (R) Incremental Linker Version 14.30.30423.0
Copyright (C) Microsoft Corporation.  All rights reserved.

/out:gh2157.exe
gh2157.obj
```

## Before PR

```
C:\Users\He\source\repos\STL\out>.\gh2157.exe
format("{:#}", 1.0)    = 1.
format("{:#.6}", 1.0)  = 1.
format("{:#g}", 1.0)   = 1.00000
format("{:#.6g}", 1.0) = 1.00000
```

## After PR

```
C:\Users\He\source\repos\STL\out>.\gh2157.exe
format("{:#}", 1.0)    = 1.
format("{:#.6}", 1.0)  = 1.00000
format("{:#g}", 1.0)   = 1.00000
format("{:#.6g}", 1.0) = 1.00000
```

Reply:
Isnt that correct? As far as I can say this is what we want.

If there is only # we want to print the digit separator but the spec says nothing about additional digits

Reply:
https://eel.is/c++draft/format.string.std#6.sentence-6 says that for `g` and `G` trailing zeroes are not removed, which implies that for _`none`_ trailing zeroes need to be removed.

Reply:
It also states
> even if no digits follow it. 
That indicates that we should not add digits.

I do not like it but i guess it would require a LWG issue

Reply:
"even if no digits follow it" doesn't indicate that we should not add digits, it only clarifies that decimal-point character should always be present.

Reply:
Thanks @statementreply, @miscco, and @AdamBucior. My interpretation:

* I believe that "After PR" is correct.
* I'd like to ping @vitaut because of the libfmt behavior difference.
* I think that there's an LWG Issue here, in the sentences that you discussed above.

`"{:#g}"` and `"{:#.6g}"` - unambiguous
---
For `"{:#g}"` and `"{:#.6g}"`, where After PR and libfmt agree that the output is `"1.00000"`, I believe that the Standardese is clear. N4892 [[format.string.std]/22][p22] passes "`6` if *precision* is not specified" so they are certainly equivalent. `to_chars` `chars_format::general, 6` wants to trim the zeros and decimal point, but [[format.string.std]/6][p6] suppresses both trimmings from happening.

`"{:#.6}"` - wording issue
---
For `"{:#.6}"`, where After PR and libfmt agree that the output is `"1.00000"`, I believe that the Standardese is technically unclear, and should be patched with an LWG Issue. According to [[format.string.std]/22][p22], this has a type of "none", but because the *precision* is specified, we call `to_chars` with `chars_format::general, 6`. As before, `to_chars` wants to trim the zeros and decimal point. [[format.string.std]/6 sentence 4][p6s4] definitely keeps the decimal point. However, [[format.string.std]/6 sentence 6][p6s6] keeps trailing zeros "for `g` and `G` conversions".

:grey_question: The question is, because type-none-with-precision performs a `chars_format::general, 6` conversion, should it be considered a `g` conversion?

:grey_exclamation: I believe the answer is "yes" - the user's mental model should be "if I pass a precision but not a letter, the letter defaults to `g` and I get all the behavior that I would have gotten if I explicitly said `g`". This is true at the `to_chars` layer (as the call is exactly `chars_format::general, 6`) and the same should be true for any `format` post-processing/modified behavior. The fact that libfmt displays this behavior is some additional evidence in favor of this being a reasonable interpretation.

`"{:#}"` - implementation divergence
---
For `"{:#}"`, where After PR outputs `"1."`, these sentences are even more important. [[format.string.std]/22][p22] clearly begins by calling `to_chars(first, last, value)`, i.e. this is plain shortest, not general precision. Thus, unlike `"{:#.6}"` which (in my opinion above) should be considered an "honorary" `g` conversion, `"{:#}"` should not be considered a `g` conversion. (Plain shortest switches between fixed and scientific with the fewest-characters, tiebreak-prefers-fixed criterion; it does not trim zeros because it doesn't emit them in the first place.) Again, [[format.string.std]/6 sentence 4][p6s4] definitely keeps the decimal point.

:question: The question here is, does [[format.string.std]/6 sentence 6][p6s6] apply?

:exclamation: I believe the answer is no: this is not a `g` conversion at the `format` specifier level, this is not a `chars_format::general` conversion at the `to_chars` level, no precision was passed to `to_chars`, and no trailing zeros were removed by `to_chars`. (There's definitely no "default of 6 precision" involved here, as plain shortest is information-preserving - it may need to emit up to 17 significant digits to round-trip a `double`.)

**Proposed resolution:** :smile_cat:
---
I believe that this could be clarified in the Standard by changing [[format.string.std]/6 sentence 6][p6s6] from:

> In addition, for `g` and `G` conversions, trailing zeros are not removed from the result.

to:

> In addition, for conversions that call `to_chars` with `chars_format::general`, trailing zeros are not removed from the result.

[p6]: https://eel.is/c++draft/format.string.std#6
[p6s4]: https://eel.is/c++draft/format.string.std#6.sentence-4
[p6s6]: https://eel.is/c++draft/format.string.std#6.sentence-6
[p22]: https://eel.is/c++draft/format.string.std#22

Reply:
I've filed an **LWG issue needed** issue, #2192.

Reply:
Great analysis, @StephanTLavavej. All makes sense and even though I'm not a big fan of `1.` it is consistent with my interpretation of the wording and {fmt} should probably do the same in this case.

Reply:
We could also change the wording to add a trailing zero if the dot is added